server/channels/app/post.go

```diff
@@ -2963,122 +3027,399 @@ func getRewritePromptForAction(action model.RewriteAction, message string, custo
-			PostID:   post.Id,
-			ExpireAt: expiryForUser,
-		}
-		_, err = a.Srv().Store().ReadReceipt().Save(rctx, receipt)
-		if err != nil {
-			return nil, model.NewAppError("RevealPost", "app.reveal_post.read_receipt.save.error", nil, "", http.StatusInternalServerError).Wrap(err)
-		}
+	// Create new read receipt for first-time reveal
+	// TODO: replace with configuration values
+	readDurationMillis := int64(model.DefaultReadDurationSeconds * 1000)
+	userExpireAt := min(postExpireAt, currentTime+readDurationMillis)
+
+	receipt = &model.ReadReceipt{
+		UserID:   userID,
+		PostID:   post.Id,
+		ExpireAt: userExpireAt,
 	}
 
-	// in case if the user trying to reveal a message that is already expired
-	if receipt.ExpireAt < now {
-		return nil, model.NewAppError("RevealPost", "app.reveal_post.read_receipt_expired.error", nil, "", http.StatusForbidden)
+	if _, err := a.Srv().Store().ReadReceipt().Save(rctx, receipt); err != nil {
+		return nil, false, model.NewAppError("RevealPost", "app.reveal_post.read_receipt.save.error", nil, "", http.StatusInternalServerError).Wrap(err)
 	}
 
-	post, appErr := a.getBurnOnReadPost(rctx, post)
-	if appErr != nil {
-		return nil, appErr
+	// If all recipients have read the post, update temporary post expiration
+	if err := a.updateTemporaryPostIfAllRead(rctx, post, receipt); err != nil {
+		// Log warning but don't fail the operation
+		rctx.Logger().Warn("Failed to update temporary post expiration after all recipients read", mlog.String("post_id", post.Id), mlog.Err(err))
+	}
+
+	return receipt, true, nil
+}
+
+// updateTemporaryPostIfAllRead updates the temporary post expiration if all recipients
+// have read the post. This ensures the post expires when the last reader's receipt expires.
+func (a *App) updateTemporaryPostIfAllRead(rctx request.CTX, post *model.Post, receipt *model.ReadReceipt) *model.AppError {
+	unreadCount, err := a.Srv().Store().ReadReceipt().GetUnreadCountForPost(rctx, post)
+	if err != nil {
+		return model.NewAppError("RevealPost", "app.reveal_post.read_receipt.get_unread_count.error", nil, "", http.StatusInternalServerError).Wrap(err)
+	}
+
+	// If there are still unread recipients, no update needed
+	if unreadCount > 0 {
+		return nil
+	}
+
+	// All recipients have read - update temporary post expiration to match receipt
+	tmpPost, err := a.Srv().Store().TemporaryPost().Get(rctx, post.Id)
+	if err != nil {
+		return model.NewAppError("RevealPost", "app.post.get_post.app_error", nil, "", http.StatusInternalServerError).Wrap(err)
+	}
+
+	tmpPost.ExpireAt = receipt.ExpireAt
+
+	if _, err := a.Srv().Store().TemporaryPost().Save(rctx, tmpPost); err != nil {
+		return model.NewAppError("RevealPost", "app.post.get_post.app_error", nil, "", http.StatusInternalServerError).Wrap(err)
+	}
+
+	return nil
+}
+
+// validateReadReceiptNotExpired ensures the user's read receipt hasn't expired.
+func (a *App) validateReadReceiptNotExpired(receipt *model.ReadReceipt, currentTime int64) *model.AppError {
```

Comment:
Nit: could combine this with the other expiration check, so that you just have one function that just takes in two timestamps.